### PR TITLE
Overhaul fuzzing

### PR DESCRIPTION
**Pull request by apoelstra**
Several big changes here:
* Moves fuzzing to its own workspace with a `contrib/test.sh` etc so that CI will check that it compiles
* FIx all warnings, clippy lints, MSRV problems, etc.; mostly move to Rust 2018
* Merge `hashes/` fuzztests into workspace
* Rewrite all scripts; add file that auto-generates CI fuzz job and Cargo.toml so we don't have to manually keep these in sync
* Remove bitrotted and partial AFL support.

Supercedes #1422 

I suspect the hashes fuzztests will actually fail since we haven't touched them in so long. Will address that if CI fails here.

### CODE REVIEW

**Comment by apoelstra** [comment]
The fuzztests are now failing because the `hashes` crate has a bunch of fuzztests which explicitly check our hash output against that of RustCrypto. However, with `cfg(fuzzing)` we have deliberately broken hashes.

For now I'll just neuter the tests. But in future I'd like to replace `cfg(fuzzing)` with a custom thing. I won't do it now since we'll need input from BlueMatt and I don't think he needs to real through this whole giant PR.

**Comment by apoelstra** [comment]
Rebase on #1733 so CI should finish faster; fix 1.41 fixes.

I think it should actually pass now :) except that I'm unsure about the `cargo fmt` fix, since this has appeared in like 5 different PRs and I'm not sure if any of them were merged..

File: fuzz/fuzz_targets/bitcoin/deserialize_witness.rs
Comment (Kixunil) [inline fuzz/fuzz_targets/bitcoin/deserialize_witness.rs:9]: Do we really need this if we're bumping MSRV to 1.48 soon anyway?

File: fuzz/contrib/test.sh
Comment (Kixunil) [inline fuzz/contrib/test.sh:34]: This shouldn't be needed unless something is seriously broken.

File: fuzz/fuzz_targets/bitcoin/script_bytes_to_asm_fmt.rs
Comment (Kixunil) [inline fuzz/fuzz_targets/bitcoin/script_bytes_to_asm_fmt.rs:29]: IIUC this could be literally `out.extend(HexIterator::new(hex).map(Result::unwrap))`

File: fuzz/fuzz-util.sh
Comment (Kixunil) [inline fuzz/fuzz-util.sh:34]: I think `nul`, `nul.*` and some other magic names are missing.

**Review by Kixunil — COMMENTED** [review]
Not-very-deep review, looks fine.

File: fuzz/fuzz-util.sh
Comment (apoelstra) [reply fuzz/fuzz-util.sh:14]: Good catch. I'll fix this, I try to be fastidious about this but I guess after too many rebases I got sloppy.

File: fuzz/contrib/test.sh
Comment (apoelstra) [reply fuzz/contrib/test.sh:34]: Heh, agreed, but I started by copying the bitcoin one and then deleted stuff, and I figured it wouldn't hurt anything to leave in.

File: fuzz/fuzz_targets/bitcoin/deserialize_witness.rs
Comment (apoelstra) [reply fuzz/fuzz_targets/bitcoin/deserialize_witness.rs:9]: Good question. No, we do not. I'll remove it and wait for #1729.

File: fuzz/fuzz_targets/bitcoin/script_bytes_to_asm_fmt.rs
Comment (apoelstra) [reply fuzz/fuzz_targets/bitcoin/script_bytes_to_asm_fmt.rs:29]: Good call. This code comes from an ancient version of rust-lightning and long predates us having any hex code :).

File: fuzz/fuzz_targets/hashes/ripemd160.rs
Comment (apoelstra) [reply fuzz/fuzz_targets/hashes/ripemd160.rs:10]: It's fuzzing that we don't panic when hashing stuff, I guess. And that we don't somehow let our engines get out of sync with our `hash` functions.

But in short, no, these fuzz tests have little value. I can can remove them entirely if you think they're too much noise.

**Comment by sanket1729** [comment]
@apoelstra, what is the motivation of bringing fuzz tests into main test directory? Does this mean we have to maintain MSRV support for hongfuzz tests?

File: fuzz/Cargo.toml
Comment (tcharding) [inline fuzz/Cargo.toml:15]: We don't need this anymore, right?

**Review by tcharding — APPROVED** [review]
Pretty cool. A few questions/comments

- Perhaps run `./generate-files.sh` again
- Why the camel case?
- Why the 2 character indentation?

And since its "pester Andrew with pointless questions" day, what column width do you use in your editor for text files? I was unable to work it out from reading the files (I've hit this problem before :)

**Comment by apoelstra** [comment]
@sanket1729 the motivation is that otherwise testing the fuzz tests involves special-case logic rather than just treating it like a worktree. No, we don't need to maintain MSRV support for the fuzztests ... but given that we currently do, and the tests themselves are all super simple, it would surprise me if this would be a problem.

@tcharding 

* `git rebase b260a310^ -x './fuzz/generate-files.sh && git diff --quiet -- fuzz/Cargo.toml` runs clean on my local version of this PR. I think the pushed version is out of date. I will push a new one once #1729 is in.
* camelCase and 2-space indents are nix habits. I can change them if they bug you.
* As for column width, I generally target 80 but often let individual lines go over if they're isolated or otherwise seem not worth breaking up. The editor doesn't enforce this, I do it by hand.

File: fuzz/Cargo.toml
Comment (apoelstra) [reply fuzz/Cargo.toml:15]: Correct. I've fixed it locally but haven't pushed, sorry.

**Comment by sanket1729** [comment]
@tcharding, I suspect that you unintentionally edited Andrew's comment above to remove his responses to me?

**Comment by tcharding** [comment]
Oh bother, how on earth did I do that. Sorry man.

**Comment by tcharding** [comment]
> * `git rebase b260a310^ -x './fuzz/generate-files.sh && git diff --quiet -- fuzz/Cargo.toml` runs clean on my local version of this PR. I think the pushed version is out of date. I will push a new one once #1729 is in.

Sweet.

> * camelCase and 2-space indents are nix habits. I can change them if they bug you.

Its ok, no need for total uniformity in shell scripts, I was just wondering where it came from.

> * As for column width, I generally target 80 but often let individual lines go over if they're isolated or otherwise seem not worth breaking up. The editor doesn't enforce this, I do it by hand.

Ah that would explain it. Cheers.

**Comment by tcharding** [comment]
There was no revert but the comment had a history of edits so I cut'n'pasted the content from @apoelstra last edit. Thanks for flagging it @sanket1729

**Comment by tcharding** [comment]
I've been getting in the habit of editing peoples post to cut'n'paste so as to get access to all the markdown, perhaps I messed up while doing that, maybe its a bad habit.

**Comment by apoelstra** [comment]
@tcharding if you select the relevant text, *then* click the ... button then "Quote Reply", Github will auto-fill your thing with just the text you selected, markdown included.

Not sure how to quote multiple people this way.

Lol @ you using maintainer powers to just get the text content...it never occurred to me I could do that.

**Comment by tcharding** [comment]
> Lol @ you using maintainer powers to just get the text content...it never occurred to me I could do that.

mmmm abuse of power :)

**Comment by tcharding** [comment]
> if you select the relevant text, _then_ click the ... button then "Quote Reply", Github will auto-fill your thing with just the text you selected, markdown included.

That is sick. After many years I'm finally starting to warm to GitHub.

**Review by tcharding — DISMISSED** [review]
ACK 13b5695c296eb8f60e31cf81bc7658df42c43346

**Comment by tcharding** [comment]
Because I always pick new contributors up on it so it feels remiss of me not to hold you to the same standards; commit `0257ed82 fuzz: ci: remove incorrrect ".rs" from CI` should have been part of the commit before (<that emoji with the tounge sticking out>)

**Comment by sanket1729** [comment]
Because this allows the fuzzer to be run at the crate root, the hfuzz files are created at crate root. Can you also them to .gitignore?

**Comment by sanket1729** [comment]
```diff
diff --git a/.gitignore b/.gitignore
index 2ccfc05f..a70f2985 100644
--- a/.gitignore
+++ b/.gitignore
@@ -20,3 +20,5 @@ bitcoin/fuzz/hfuzz_target
 bitcoin/fuzz/hfuzz_workspace
 hashes/fuzz/hfuzz_target
 hashes/fuzz/hfuzz_workspace
+hfuzz_workspace
+hfuzz_target

```

**Comment by apoelstra** [comment]
Squashed the two commits @tcharding suggested and updated `.gitignore` as @sanket1729 suggseted.

**Review by tcharding — APPROVED** [review]
ACK 38569cb237ee9c50b0b63ddf8463d30a71421001

**Comment by tcharding** [comment]
Thanks man.

**Comment by apoelstra** [comment]
@sanket1729 @Kixunil do either of you have outstanding problems with this PR?

**Comment by apoelstra** [comment]
Rebased.

**Comment by apoelstra** [comment]
What is holding up this PR? I would like to remove the `cfg(fuzzing)` junk from `hashes` (or at least, move it to its own cfg flag) but want to get this in first.

**Review by tcharding — DISMISSED** [review]
ACK 59d7ab6dc913ac9951da82fbc6a35d3770f5649a

**Comment by sanket1729** [comment]
Reviewing this again today. But looks like something else broke in embedded world

**Comment by apoelstra** [comment]
Rebased on `master`.

**Review by tcharding — APPROVED** [review]
ACK 2860aae1a5ded81d96af14b5df0d9a8833a4fd8f

**Comment by apoelstra** [comment]
@sanket1729 can you ACK using the github api? it looks like you just commented.

**Review by sanket1729 — APPROVED** [review]

**Comment by apoelstra** [comment]
Thanks!

**Pull request merged**
